src/webui/api/torrentscontroller.cpp

```diff
@@ -1523,3 +1661,204 @@ void TorrentsController::setSSLParametersAction()
 
     torrent->setSSLParameters(sslParams);
 }
+
+void TorrentsController::metadataAction()
+{
+    const QStringList sources = params()[u"sources"_s].split(u',', Qt::SkipEmptyParts);
+    const DataMap &uploadedTorrents = data();
```

Comment:
I strongly believe that retrieving metadata by magnet URI (or info hash) and parsing given torrent file are different kinds of jobs. It is very bad idea to develop such a "crazy combine" finctions. It makes unreadable and unmaintainable both the function itself and the client code that analyzes its result.

Reply:
It seems you're not familiar with `TorrentsController::addAction()`. And I strongly disagree that the code is either unreadable or unmaintainable because of this.

Reply:
> It seems you're not familiar with `TorrentsController::addAction()`.

Judging by your intentions and questions like above one, you do not think of this as a separate independent function, but as a closely tied addition to `torrents/add` function, the semantics of which I would describe with the name "prepare metadata", which is intended specifically to precede the call to `torrents/add`.

Reply:
Well, if everyone is satisfied with this form of existence of interdependent functions (endpoints), so be it. But I would still expect it to have a name that more clearly reflects its purpose. Just "metadata" is too vague.

@Chocobo1, have you something to comment on it?

Reply:
> Although, by providing metadata in the form it does now, it is hardly suitable to be a general purpose function.

How so? This API provides metadata associated with the torrent. It does not do so in any format that's specific to the `torrents/add`. API. There does exist a cache that's shared between the APIs, but that is purely an implementation detail.

Reply:
But I still don't see the point to use it to get the metadata of an already added torrent in such a form it does.

Reply:
> How so? This API provides metadata associated with the torrent. It does not do so in any format that's specific to the `torrents/add`. API. There does exist a cache that's shared between the APIs, but that is purely an implementation detail.

If it's supposed to be an independent function after all, why are you "rubbing my nose" at `TorrentsController::addAction()` (in your comment above)?

Reply:
> But I still don't see the point to use it to get the metadata of an already added torrent in such a form it does.

What do you mean - I don't understand? Metadata is returned in the same format regardless of whether the torrent is already added.

> If it's supposed to be an independent function after all, why are you "rubbing my nose" at TorrentsController::addAction() (in your comment above)?

I think you're intentionally misreading what I wrote. I'll directly copy and paste what I wrote [below](https://github.com/qbittorrent/qBittorrent/pull/21015#issuecomment-2231267749):

> My point in bringing up `TorrentsController::addAction()` is to make clear that these are existing patterns and not ones that I invented myself.